src/java.base/share/classes/java/lang/ClassValue.java

```diff
@@ -220,29 +212,52 @@ private T getFromBackup(Entry<?>[] cache, Class<?> type) {
     private T getFromHashMap(Class<?> type) {
         // The fail-safe recovery is to fall back to the underlying classValueMap.
         ClassValueMap map = getMap(type);
-        for (;;) {
-            Entry<T> e = map.startEntry(this);
-            if (!e.isPromise())
-                return e.value();
+        var accessed = map.readAccess(this);
+        if (accessed instanceof Entry) {
+            @SuppressWarnings("unchecked")
+            var cast = (Entry<T>) accessed;
+            return cast.value;
+        }
+
+        RemovalToken permission = (RemovalToken) accessed; // nullable
```

Comment:
This might be better renamed to `removalToken` or just `token` instead of `permission`, in order to have a consistent terminology.